File: Erlang/erlangexamples/apps/erlangexamples/src/closures.erl
%%%-------------------------------------------------------------------
%%% @author Wolfgang Loder
%%% @doc
%%% Erlang and Elixir for Imperative Programmers, 2016
%%% Chapter 14 - Closures
%%% @end
%%%-------------------------------------------------------------------
-module(closures).
-author("Wolfgang Loder").

-export ([print_closure/0]).

initiate_closure() ->
    [tomato,onion,cheese].

print_closure() ->
    L = initiate_closure(),
    lists:map(fun(A)-> io:format("~p~n", [A]) end, L).

File: Elixir/elixirexamples/closures.ex
@moduledoc """
Author: Wolfgang Loder
Email: wolfgang.loder@googlemail.com
Date: June 2016
Book Erlang and Elixir for Imperative Programmers, Chapter 14
"""
@doc """
Clojure Examples
"""

defmodule ClosuresExample do
    def initiate_closure do
        [:tomato,:onion,:cheese]
            |> Enum.map &(fn () -> &1 end)
    end

    def print_closure do
        initiate_closure
            |> Enum.each &(IO.puts(&1.()))
    end
end
